### PR TITLE
Fix tramp crash

### LINKED ISSUE #139: args-out-of-range error when using TRAMP
When editing a remote file over TRAMP, if `annotate-use-echo-area` is set to `nil`, I get the following error:
```
Debugger entered--Lisp error: (args-out-of-range "52 53")
  signal(args-out-of-range ("52 53"))
  tramp-error((tramp-file-name "ssh" nil nil "desktop" nil "/home/gsgx/code/pwndbg/README.md" nil) args-out-of-range "52 53")
  tramp-signal-hook-function(args-out-of-range (52 53))
  remove-text-properties(52 53 (display nil))
  (save-excursion (goto-char end) (end-of-line) (remove-text-properties (point) (1+ (point)) '(display nil)))
  (let ((saved-undo-list (copy-tree buffer-undo-list t))) (buffer-disable-undo) (save-excursion (goto-char end) (end-of-line) (remove-text-properties (point) (1+ (point)) '(display nil))) (setq buffer-undo-list saved-undo-list) (buffer-enable-undo))
  (let ((read-mode-p (if buffer-read-only 1 -1))) (if (= read-mode-p 1) (progn (read-only-mode -1))) (let ((saved-undo-list (copy-tree buffer-undo-list t))) (buffer-disable-undo) (save-excursion (goto-char end) (end-of-line) (remove-text-properties (point) (1+ (point)) '(display nil))) (setq buffer-undo-list saved-undo-list) (buffer-enable-undo)) (if (= read-mode-p 1) (progn (read-only-mode 1))))
  (progn (let ((read-mode-p (if buffer-read-only 1 -1))) (if (= read-mode-p 1) (progn (read-only-mode -1))) (let ((saved-undo-list (copy-tree buffer-undo-list t))) (buffer-disable-undo) (save-excursion (goto-char end) (end-of-line) (remove-text-properties (point) (1+ (point)) '(display nil))) (setq buffer-undo-list saved-undo-list) (buffer-enable-undo)) (if (= read-mode-p 1) (progn (read-only-mode 1)))))
  (unwind-protect (progn (let ((read-mode-p (if buffer-read-only 1 -1))) (if (= read-mode-p 1) (progn (read-only-mode -1))) (let ((saved-undo-list (copy-tree buffer-undo-list t))) (buffer-disable-undo) (save-excursion (goto-char end) (end-of-line) (remove-text-properties (point) (1+ (point)) '(display nil))) (setq buffer-undo-list saved-undo-list) (buffer-enable-undo)) (if (= read-mode-p 1) (progn (read-only-mode 1))))) (if modified nil (restore-buffer-modified-p nil)))
  (let* ((modified (buffer-modified-p)) (buffer-undo-list t) (inhibit-read-only t) (inhibit-modification-hooks t)) (unwind-protect (progn (let ((read-mode-p (if buffer-read-only 1 -1))) (if (= read-mode-p 1) (progn (read-only-mode -1))) (let ((saved-undo-list (copy-tree buffer-undo-list t))) (buffer-disable-undo) (save-excursion (goto-char end) (end-of-line) (remove-text-properties (point) (1+ ...) '...)) (setq buffer-undo-list saved-undo-list) (buffer-enable-undo)) (if (= read-mode-p 1) (progn (read-only-mode 1))))) (if modified nil (restore-buffer-modified-p nil))))
  (progn (message "%d" (buffer-size)) (message "%s" (buffer-name)) (message "%s" (buffer-substring (point-min) (point-max))) (let* ((modified (buffer-modified-p)) (buffer-undo-list t) (inhibit-read-only t) (inhibit-modification-hooks t)) (unwind-protect (progn (let ((read-mode-p (if buffer-read-only 1 -1))) (if (= read-mode-p 1) (progn (read-only-mode -1))) (let ((saved-undo-list ...)) (buffer-disable-undo) (save-excursion (goto-char end) (end-of-line) (remove-text-properties ... ... ...)) (setq buffer-undo-list saved-undo-list) (buffer-enable-undo)) (if (= read-mode-p 1) (progn (read-only-mode 1))))) (if modified nil (restore-buffer-modified-p nil)))))
  (if (and (> (buffer-size) 0) (not (buffer-narrowed-p))) (progn (message "%d" (buffer-size)) (message "%s" (buffer-name)) (message "%s" (buffer-substring (point-min) (point-max))) (let* ((modified (buffer-modified-p)) (buffer-undo-list t) (inhibit-read-only t) (inhibit-modification-hooks t)) (unwind-protect (progn (let ((read-mode-p ...)) (if (= read-mode-p 1) (progn ...)) (let (...) (buffer-disable-undo) (save-excursion ... ... ...) (setq buffer-undo-list saved-undo-list) (buffer-enable-undo)) (if (= read-mode-p 1) (progn ...)))) (if modified nil (restore-buffer-modified-p nil))))))
  annotate--remove-annotation-property(1 52)
  internal-default-process-filter(#<process *tramp/ssh desktop*> "are you awake\n///48b560667ab24e58172a75e839e9314a#...")
  accept-process-output(#<process *tramp/ssh desktop*> nil nil t)
  tramp-accept-process-output(#<process *tramp/ssh desktop*>)
  tramp-wait-for-regexp(#<process *tramp/ssh desktop*> 10 "\\(^\\|\0\\)[^#$\n]*///48b560667ab24e58172a75e839e9314a...")
  tramp-wait-for-output(#<process *tramp/ssh desktop*> 10)
  ...
  self-insert-command(1 32)
  funcall-interactively(self-insert-command 1 32)
  call-interactively(self-insert-command nil nil)
  command-execute(self-insert-command)
  ```
  
This does not happen when `annotate-use-echo-area` is `t`. After some debugging, I found out that if I removed `annotate--change-guard` call from this code, the error goes away.
```elisp
    (if annotate-use-echo-area
      (font-lock-add-keywords
       nil
       '((annotate--font-lock-matcher (2 (annotate--annotation-builder)))))
    (font-lock-add-keywords
     nil
     '((annotate--font-lock-matcher (2 (annotate--annotation-builder))
                                    (1 (annotate--change-guard)))))))   ;; remove this call
```

I also made this modification to help debug the issue:
```elisp
(defun annotate--remove-annotation-property (_begin end)
  "Cleans up annotation properties associated within a region
surrounded by `BEGIN' and `END'."
  (when (and (> (buffer-size) 0)
             (not (buffer-narrowed-p)))
    (message "%d" (buffer-size))
    (message "%s" (buffer-name))
    (message "%s" (buffer-substring (point-min) (point-max)))
    ...
 ```
 Look at the messages buffer now, I see:
 ```
 57
*tramp/ssh desktop*
tramp_exit_status 0
///48b560667ab24e58172a75e839e9314a#$
Args out of range: "Args out of range", "58 59"
```

Note that the name of the buffer is `*tramp/ssh desktop*`, which is a buffer created by TRAMP, it's not the buffer I attempted to edit.

Do you have any idea why this might be happening?

### PR DESCRIPTION
**Pull request by cage2**
Hi!

This patch should fix #139, it is a bit a workaround but i the best i was able to figure out. :(

Bye!
C.

### CODE REVIEW

**Comment by cage2** [comment]
Hi @bastibe !

Do you think we can merge this branch?

Bye!
C.

File: annotate.el [hunk 1]
```diff
@@ -2482,8 +2482,7 @@ sophisticated way than plain text."
                                       (separator-re    "\^L?\^_\^L?\^J")
                                       (has-separator-p (string-match separator-re file-contents))
                                       (has-node-p      (string-match "Node:" file-contents)))
-                                 (if (or (annotate-info-root-dir-p filename)
-                                         (and has-separator-p
+                                 (if (or (and has-separator-p
```
Comment (bastibe) [inline annotate.el]: The logic of `((has-sep && has-node) || (has-sep && has-info))` could be compacted to `(has-sep && (has-node || has-info))`.

**Review by bastibe — COMMENTED** [review]
Hey, looks good! Sorry for not responding for so long, it's been a busy time.

File: annotate.el (hunk 1 above)
Comment (cage2) [reply annotate.el]: Thank @bastibe !

Your help is precious as usual!

Fixed! :)

Bye!
C.

**Comment by bastibe** [comment]
Great! Feel free to merge.

By the way, it might be useful to think about setting up automated tests. I have actually no idea how to do that in Emacs, but I found it a great help in other projects.

**Comment by cage2** [comment]
Hi @bastibe !

> Great! Feel free to merge.

Thanks for your help!
 
> By the way, it might be useful to think about setting up automated tests. I have actually no idea how to do that in Emacs, but I found it a great help in other projects.

I also have no clue how to start to do testing in Emacs but nevertheless I think your is a good Idea! :)
I am adding a issue until i figure out how to write the test; i think i would  start considering [ERT](https://www.gnu.org/software/emacs/manual/html_node/ert/index.html), the library that is included in Emacs.

Merging in a seconds!
Bye!
C.

**Pull request merged**